### PR TITLE
Stream FormExportInstances to Angular

### PR DESCRIPTION
**Pull request by kaapstorm**
Context: https://manage.dimagi.com/default.asp?274447
TL;DR: A project space is running out of memory when Angular fetches a list of FormExportInstances.

In this PR:
* Iterate FormExportInstances with `iter_docs`
* A generator that yields a JSON document and allows generators in the input data. (At the moment we only need dictionaries, but trivial (and tempting) to make more generic.)
* A StreamingHttpResponse subclass to use this JSON generator

I'm opening this PR for some feedback on my approach. If this seems reasonable, I'll add tests for `FormExportListView` and test this properly. If I've duplicated a similar approach somewhere else, I'd be happy to reuse existing work, or adapt to an existing pattern/idiom.

Might be easiest to :blowfish: but probably small enough to :office:.

### CODE REVIEW

**Comment by emord** [comment]
It does seem better to use `iter_docs`, but maybe this page should be paginated in some way?

**Comment by dannyroberts** [comment]
I'm okay with trying this out, but my guess is that anything that we can't hold as JSON in memory on the server, angular's also not going to be able to handle gracefully, since on the broswer side, it _will_ have to hold it all in memory. Not sure what you think, but thought I'd throw that out there.

Might have to merge with master to get tests to pass.

**Comment by kaapstorm** [comment]
@dannyroberts yes, I confess to knowingly making it the user's problem, in the hope that they have more memory to spend on one user's request than the server does. :grimacing: ... but totally agree with @emord -- pagination is a very good idea.

**Review by snopoke — COMMENTED** [review]
Not sure how I feel about this. As @dannyroberts points out this might solve the server issue but likely just push the issue to the browser. 

This also adds a fair amount of complexity that will be unnecessary if we paginate the view which is the better long term solution.

I also don't think the change to paginate the view will be much more effort that what was required for this change (though could be wrong)

**Comment by kaapstorm** [comment]
I think being able to stream JSON could be really useful, but I agree that paginating is the right approach here, for the UI, and for the browser.

If we do find a usecase for streaming results, most of the complexity that this PR adds is because we've got two ways of fetching the same data. That can be eliminated by switching to using generators instead of lists. I don't think there are many reasons to traverse a resultset more than once if sorting and counting are done by the database. (Sorting might need to be a longer (offline) discussion.)

**Comment by kaapstorm** [comment]
Closing, but I'll keep the branch around, for now.

**Comment by biyeun** [comment]
@snopoke the issue with pagination comes with the sorting and filtering that we do to the records after they are fetched. so it's not that simple of a task

**Comment by biyeun** [comment]
gonna merge this to urgently unblock a partner once tests pass, but @kaapstorm let's discuss tackling this with a more solid solution. I got to the point where I needed to tweak the couch view to do it properly and went uh oh...this is not a teeny change. lol
